# Week 3 — Imbalanced Classification & LightGBM Modeling
## Member 3 — Context & Integration

## 1. Environment Setup & Imports
## 2. Load Fused Dataset
## 3. Feature Matrix Extraction
## 4. NaN Verification
## 5. Class Distribution

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.preprocessing import LabelEncoder

print("All imports successful")

All imports successful


## 2. Load Fused Dataset

In [3]:
df = pd.read_csv('../data/ai4i2020.csv')

np.random.seed(42)
df['ambient_temp_C'] = np.random.normal(loc=28, scale=5, size=len(df))
df['factory_load_pct'] = np.random.uniform(50, 100, size=len(df))
df['humidity_pct'] = np.random.normal(loc=60, scale=10, size=len(df))

le = LabelEncoder()
df['Type_enc'] = le.fit_transform(df['Type'])

print("Fused dataset loaded!")
print("Shape:", df.shape)

Fused dataset loaded!
Shape: (10000, 18)


In [4]:
ext_features = [
    'Air temperature [K]', 'Process temperature [K]',
    'Rotational speed [rpm]', 'Torque [Nm]', 'Tool wear [min]',
    'Type_enc', 'ambient_temp_C', 'factory_load_pct', 'humidity_pct'
]

X = df[ext_features]
y = df['Machine failure']

print("Feature Matrix X shape:", X.shape)
print("Target y shape:", y.shape)

Feature Matrix X shape: (10000, 9)
Target y shape: (10000,)


In [5]:
print("NaN check:")
print(X.isnull().sum().sum())

if X.isnull().sum().sum() == 0:
    print("✅ No NaN values — dataset is clean!")

NaN check:
0
✅ No NaN values — dataset is clean!


In [6]:
print("Class Distribution:")
print(y.value_counts())
print(f"\nNo Failure: {(y==0).sum()} — {(y==0).mean()*100:.2f}%")
print(f"Failure: {(y==1).sum()} — {(y==1).mean()*100:.2f}%")
print(f"Imbalance Ratio: {(y==0).sum()/(y==1).sum():.1f}:1")
print("\n✅ Dataset ready for SMOTE + LightGBM modeling!")

Class Distribution:
Machine failure
0    9661
1     339
Name: count, dtype: int64

No Failure: 9661 — 96.61%
Failure: 339 — 3.39%
Imbalance Ratio: 28.5:1

✅ Dataset ready for SMOTE + LightGBM modeling!


## Commentary — Feature Matrix & Class Distribution

- Total features: 9 (5 internal sensors + Type_enc + 3 external context)
- Total samples: 10,000
- No NaN values — dataset clean and ready
- Imbalance Ratio: 28.5:1 — SMOTE will be applied inside training folds only